utils/wasm-instrument/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+pub const GLOBAL_NAME_ALLOWANCE: &str = "gear_allowance";
+
+pub const IMPORT_NAME_OUT_OF_GAS: &str = "gr_out_of_gas";
+pub const IMPORT_NAME_OUT_OF_ALLOWANCE: &str = "gr_out_of_allowance";
+
+pub fn inject<R: Rules>(
+    module: elements::Module,
+    rules: &R,
+    gas_module_name: &str,
+) -> Result<elements::Module, elements::Module> {
+    if module
+        .import_section()
+        .map(|section| {
+            section.entries().iter().any(|entry| {
+                entry.module() == gas_module_name
+                    && (entry.field() == IMPORT_NAME_OUT_OF_GAS
+                        || entry.field() == IMPORT_NAME_OUT_OF_ALLOWANCE)
+            })
+        })
+        .unwrap_or(false)
+    {
+        return Err(module);
+    }
+
+    if module
+        .export_section()
+        .map(|section| {
+            section.entries().iter().any(|entry| {
+                entry.field() == GLOBAL_NAME_ALLOWANCE || entry.field() == GLOBAL_NAME_GAS
+            })
+        })
+        .unwrap_or(false)
+    {
+        return Err(module);
+    }
+
+    let mut mbuilder = builder::from_module(module);
+
+    // fn out_of_...() -> ();
+    let import_sig = mbuilder.push_signature(builder::signature().build_sig());
+
+    mbuilder.push_import(
+        builder::import()
+            .module(gas_module_name)
+            .field(IMPORT_NAME_OUT_OF_GAS)
+            .external()
+            .func(import_sig)
+            .build(),
+    );
+
+    mbuilder.push_import(
+        builder::import()
+            .module(gas_module_name)
+            .field(IMPORT_NAME_OUT_OF_ALLOWANCE)
+            .external()
+            .func(import_sig)
+            .build(),
+    );
+
+    // back to plain module
+    let module = mbuilder.build();
+
+    let import_count = module.import_count(elements::ImportCountType::Function);
+    let out_of_gas_index = import_count as u32 - 2;
+    let out_of_allowance_index = import_count as u32 - 1;
+
+    let gas_charge_index = module.functions_space();
+    let gas_index = module.globals_space() as u32;
+    let allowance_index = gas_index + 1;
+
+    let mut mbuilder = builder::from_module(module);
+
+    mbuilder.push_global(
+        builder::global()
+            .value_type()
+            .i64()
+            .init_expr(Instruction::I64Const(0))
+            .mutable()
+            .build(),
+    );
+
+    mbuilder.push_export(
+        builder::export()
+            .field(GLOBAL_NAME_GAS)
+            .internal()
+            .global(gas_index)
+            .build(),
+    );
+
+    mbuilder.push_global(
+        builder::global()
+            .value_type()
+            .i64()
+            .init_expr(Instruction::I64Const(0))
+            .mutable()
+            .build(),
+    );
+
+    mbuilder.push_export(
+        builder::export()
+            .field(GLOBAL_NAME_ALLOWANCE)
+            .internal()
+            .global(allowance_index)
+            .build(),
+    );
+
+    let mut elements = vec![
```

Comment:
Create TODO and isssue to optimize this code

Reply:
ok